lib/bolt/inventory/inventory2.rb

```diff
@@ -304,17 +272,141 @@ def inherit_data(track, name, data)
+      def self.parse_uri(string)
+        # TODO: is this OK, breaking gem loading assertions here
+        require 'addressable/uri'
+        if string.nil?
+          nil
+        # Forbid empty uri
+        elsif string.empty?
+          raise Bolt::ParseError, "Could not parse target URI: URI is empty string"
+        elsif string =~ %r{^[^:]+://}
+          Addressable::URI.parse(string)
+        else
+          # Initialize with an empty scheme to ensure we parse the hostname correctly
+          Addressable::URI.parse("//#{string}")
+        end
+      rescue Addressable::URI::InvalidURIError => e
+        raise Bolt::ParseError, "Could not parse target URI: #{e.message}"
+      end
+
+      # I guess these dont really belong with Target
+      def set_var(target, var_hash)
+        @targets[target.name]['vars'] = @targets[target.name]['vars'].merge(var_hash)
+      end
+
+      def vars(target)
+        @targets[target.name]['vars']
+      end
+
+      def add_facts(target, new_facts = {})
+        # CODEREVIEW: Do we want to warn for set methods?
+        @logger.warn("No facts to add") if new_facts.empty?
+        @targets[target.name]['facts'] = Bolt::Util.deep_merge(@targets[target.name]['facts'], new_facts)
+      end
+
+      def facts(target)
+        @targets[target.name]['facts']
+      end
+
+      def set_feature(target, feature, value = true)
+        if value
+          @targets[target.name]['features'] << feature
+        else
+          @targets[target.name]['features'].delete(feature)
+        end
+      end
+
+      def features(target)
+        @targets[target.name]['features']
+      end
 
-        # If there is no name then this target was only referred to as a string.
-        uri = data['uri']
-        uri ||= target_name unless data['name']
+      def plugin_hooks(target)
+        @targets[target.name]['plugin_hooks']
+      end
 
-        Target.new(uri, name_opt)
+      def set_config(target, key_or_key_path, value)
+        config = key_or_key_path.empty? ? value : build_config_hash([key_or_key_path].flatten, value)
+        @targets[target.name]['config'] = @targets[target.name]['config'].merge(config)
+        update_target(target)
+      end
+
+      def build_config_hash(key_or_key_path, value)
+        # https://stackoverflow.com/questions/5095077/ruby-convert-array-to-nested-hash
+        key_or_key_path.reverse.inject(value) { |acc, key| { key => acc } }
+      end
+      private :build_config_hash
+
+      def update_conf(target, conf)
+        @targets[target.name]['protocol'] = conf[:transport]
+        t_conf = conf[:transports][target.transport.to_sym] || {}
+        @targets[target.name]['user'] = t_conf['user']
+        @targets[target.name]['password'] = t_conf['password']
+        @targets[target.name]['port'] = t_conf['port']
+        @targets[target.name]['host'] = t_conf['host']
+        @targets[target.name]['options'] = @targets[target.name]['options'].merge(t_conf)
```

Comment:
This seems like it should be `config` now? I'm not sure how `options` is different from `config` for `Bolt::Target2`. If they're different there's no harm leaving them, but it might be helpful to rename options

Reply:
Basically config is boiled down to a flat options hash for easy use in transports. The outstanding question I have about options is that with the current implementation options are not pruned when a target's config is updated. So for example if you change protocol, any overlapping options are updated as expected, but ones that are not used by the updated transport persist. So for example if I were to update from ssh to winrm, the `disconnect-timeout` option would be persisted in options.